agent/hcp/telemetry/gauge_store.go

```diff
@@ -0,0 +1,77 @@
+package telemetry
+
+import (
+	"context"
+	"sync"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+// gaugeStore holds last seen Gauge values for a particular metric (<name,last_value>) in the store.
+// OTEL does not currently have a synchronous Gauge instrument. Instead, it allows the registration of callbacks.
+// The callbacks are called during export, where the Gauge value must be returned.
+// This store is a workaround, which holds last seen Gauge values until the callback is called.
+type gaugeStore struct {
+	store map[string]*gaugeValue
+	mutex sync.Mutex
+}
+
+// gaugeValues are the last seen measurement for a Gauge metric, which contains a float64 value and labels.
+type gaugeValue struct {
+	Value      float64
+	Attributes []attribute.KeyValue
+}
+
+// NewGaugeStore returns an initialized empty gaugeStore.
+func NewGaugeStore() *gaugeStore {
+	return &gaugeStore{
+		store: make(map[string]*gaugeValue, 0),
+	}
+}
+
+// LoadAndDelete will read a Gauge value and delete it.
+// Once registered for a metric name, a Gauge callback will continue to execute every collection cycel.
+// We must delete the value once we have read it, to avoid repeat values being sent.
+func (g *gaugeStore) LoadAndDelete(key string) (*gaugeValue, bool) {
```

Comment:
Would the caller ever care if the value is not in the `gaugeStore`? Trying to determine if we need to return the ok bool?

Reply:
I check if there is a value to emit here:

https://github.com/hashicorp/consul/pull/17159/files#diff-2f8b324f4e5da8b8ad5221a272ca430aa377e759443fa347f33b6524906661c4R71-R73

Reply:
thanks for catching this, I've fixed it and for the otel test too.